### ISSUE TITLE
Introduce better field labeling and accessibility for screen readers

### ISSUE THREAD

**Issue by BenHenning** [google/blockly#9301]
### Check for duplicates

- [x] I have searched for similar issues before opening a new one.

### Problem

Screen reader experimentation (https://github.com/google/blockly-keyboard-experimentation/discussions/673) currently has somewhat insufficient labeling and context support for fields.

Specifically, block fields may lack type context (e.g. proper ARIA setup for drop-downs), or may not have a sufficiently clear described context for custom fields (e.g. colour picker).

### Request

Field accessibility largely comes down to three things:
1. Supporting a descriptive label for what the field actually represents along with its value.
2. Providing additional context for how the field works semantically (and provide some hint for how its editor will work).
3. Ensuring that field editors are themselves accessible.

(1) can be solved by introducing an expectation for a new getter for `Field` that can dynamically determine a good fitting label for the field. This getter can be provided in custom fields without an override, and is thus fine to introduce without custom versions or complicated NPM linking in plugins (since the getter will simply go unused outside of the screen reader experimentation work). Some specific considerations here:
- When does the field need to recompute its label and possibly update its properties? Some invariants/assumptions need to be established in the getter's or class's contract.
- Custom getters in plugins should be tracked so that they can be removed or changed if the long-term solution doesn't end up using them in the same way.
- How should the getter be defaulted? This should largely be on a class-by-class basis, but the specific field value may need to be considered as well.

(2) needs to be solved by introducing correct ARIA roles and, in some cases, elements on a field-by-field basis. #8205 has additional details for implementing ARIA for different fields. This is tracked by the following issues (may not be comprehensive):
- #8206
- #8210
- #8255
- #8211
- #8254
- #8212
- #8213

(3) needs to be done, at minimum, for any external plugins that the keyboard navigation plugin playground uses, plus for all built in fields. The latter are covered by the issues mentioned in (2), and the former are covered as follows:
- To be done in a follow-up issue as discovered.

### Alternatives considered

One alternative considered was introducing a utility that can determine a label based on the specific class, though this is a dependency inversion and is largely tricky to manage (even in an experiment) vs. using a more inheritance-based approach (which in this case is a bit easier to work with).

### Additional context

This solution was largely determined in a team discussion this week with contributions from essentially the whole team.

**Comment by BenHenning** [comment]
Outside of the obvious roles & labels that are being tracked by the child items of this issue, the main two gaps to figure out for fields are:
- Field editors (which ends up being very domain-specific as to how to make them accessible).
- How to contextualize field "types" (such as for the draw emoji case).

Fundamentally, this is "how do we read out field values" and "how do we edit field values" but generically. The latter can be done independent of the ongoing experimental screen reader work, but the former may need some slightly hacky workarounds to get to a reasonable state in the interim.

Perhaps we could introduce a temporary manager class that provides a label for a given field by considering its class type, custom type information (insomuch as we can introspect that for fields outside core), and the current value of the field in order to compute a label. Ideally this would be decentralized and data-driven as part of the long-term solution, but it could be workable for interim experimentation (especially if designed with extensibility in mind so that it's easy for MakeCode or micro:bit to register additional field types for better readout).

**Comment by BenHenning** [comment]
I still need to audit the external field editors that need updating, but I have an idea for contextualizing based on #9305: we can introduce two methods in `Field`:
- `getValueAriaLabel`: only returns a readable representation of the field's value. This is to be used for the block label construction in #9299.
- `getFieldAriaLabel`: returns a fully contextualized readout of the field within the context of a specific block, including the field's value. This one is more likely to change over time as we work toward a tested and finalized labeling strategy, but we currently have additional context beyond a field's value and we will likely want to maintain that customization flexibility during testing. This will also probably leverage block definition customization as established in #9305.

These are obviously not perfect names, but they're reasonable for experimentation.

**Comment by BenHenning** [comment]
So going the other direction per https://github.com/google/blockly/pull/9384#discussion_r2395629363 I think we probably actually want to just provide the value and skip the extra contextualization. Reasons:
1. Contextualizing fields is hard. We don't have good fallback details for non-custom cases (unlike blocks).
2. We're not sure we actually need the additional information, so while we're at a stage where we have to decide we may as well err on the "less effort" approach until we get user feedback suggestion otherwise.
3. The general movement for screen reader label design right now is less information when we're not sure about the context being shared, and I think additional field contexts fall in that.

**Comment by BenHenning** [comment]
Addressed in #9446.
